src/core/observable.ts

```diff
@@ -1,10 +1,58 @@
-import {IDerivation, notifyDependencyReady, notifyDependencyStale} from "./derivation";
+import {IDerivation, changeDependenciesState} from "./derivation";
 import {globalState} from "./globalstate";
-import {SimpleSet} from "../utils/set";
+
+export class DerivationsSets {
+	size: number = 0;
+	data0 = {};
+	data1 = {};
+	data2 = {};
+
+	get length() {
+		return this.size;
+	}
+	asArray() {
+		const res = new Array(this.size);
+		let i = 0;
+		for (let key in this.data0) {
+			res[i] = this.data0[key];
+			i++;
+		}
+		for (let key in this.data1) {
+			res[i] = this.data1[key];
+			i++;
+		}
+		for (let key in this.data2) {
+			res[i] = this.data2[key];
+			i++;
+		}
+		return res;
+	}
+	add(value: IDerivation) {
+		const data = this["data" + value.dependenciesState];
+		const m = value.__mapid;
+		if (!(m in data)) {
+			data[m] = value;
+			this.size++;
+		}
+	}
+	remove(value: IDerivation) {
+		const data = this["data" + value.dependenciesState];
+		const m = value.__mapid;
+		if (m in data) {
+			delete data[m];
+			this.size--;
+		}
+	}
+	// move(targetState: number, value: IDerivation) {
```

Comment:
+1 :) although if all is gathered in one set, the operation isn't needed at all. But not sure whether that is faster or terribly slow, let's investigate